### PR TITLE
[AutoDiff] @_alwaysEmitIntoClient registration tests

### PR DESCRIPTION
**Pull request by marcrasi**
Here are some negative tests demonstrating the problem in TF-1103.

What's happening is:
* When the differentiation pass is processing `gradient(at: 0, in: f)`, it finds the differentiability config arising from the `@derivative(of: f)`, and creates a corresponding differentiability witness declaration.
* The linker tries to link references to that witness to an actual definition.
* There is no actual definition because the witness linkage is PublicNonABI, which means that the witness does not get emitted into the library.

Some possible fixes are:
* Make it so that the differentiability witness is Public instead of PublicNonABI. An obstacle for this approach is that TBDGen currently bases its decision on whether to emit a TBD entry on the linkage of the original function. So TBD doesn't won't emit a TBD entry for the witness for a PublicNonABI function. We could overcome this obstacle by teaching TBDGen to use the linkage of the derivative function instead of the linkage of the original function.
* Teach the differentiation pass about PublicNonABI witnesses. Specifically:
  - Put information about the witness linkage in the `DerivativeFunctionConfigurationList`.
  - Make the pass create a decl with the appropriate linkage based on that information.
  - Teach something to deserialize these witnesses. (Maybe SIL/Linker.cpp is the correct place to do this, if the linking pass runs after differentiation. However, if the linking pass runs before differentiation, maybe we'll need the differentiation pass to deserialize these).

My initial thought is that the second approach is better because we want the visibility/linkage/etc of the derivative functions to always be as close to the original functions as possible.

### CODE REVIEW

**Review by dan-zheng — APPROVED** [review]
Nice tests!

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

File: test/AutoDiff/downstream/always_emit_into_client_multi_file.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// TODO(TF-1103): Fix this test so that there is not a linker error. Then, move this test to
+// cross_module_derivative_attr_e2e.swift.
```
Comment (compnerd) [inline test/AutoDiff/downstream/always_emit_into_client_multi_file.swift]: I believe that this would belong in the validation tests directory

Comment (marcrasi) [reply test/AutoDiff/downstream/always_emit_into_client_multi_file.swift]: What tests should go there? We haven't put any autodiff tests there -- we've put them all under `tests/`, even ones that build modules like this one.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Comment by ematejska** [comment]
@swift-ci please test tensorflow macos

**Comment by ematejska** [comment]
@swift-ci please test tensorflow clean linux

**Comment by ematejska** [comment]
@swift-ci please clean test tensorflow linux

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Pull request merged**